### PR TITLE
FEATURE: Use group SMTP settings for sending user notification emails (initial)

### PR DESCRIPTION
**Pull request by martin-brennan**
This PR changes the `UserNotification` class to send outbound `user_private_message` using the group's SMTP settings, but only if:

* The first allowed_group on the topic has SMTP configured and enabled
* SiteSetting.enable_smtp is true
* The group does not have IMAP enabled, if this is enabled the `GroupSMTPMailer` handles things

The email is sent using the group's `email_username` as both the `from` and `reply-to` address, so when the user replies from their email it will go through the group's SMTP inbox, which needs to have email forwarding set up to send the message on to a location (such as a hosted site email address like meta@discoursemail.com) where it can be POSTed into discourse's handle_mail route.

Also includes a fix to `EmailReceiver#group_incoming_emails_regex` to include the `group.email_username` so the group does not get a staged user created and invited to the topic (which was a problem for IMAP), as well as updating `Group.find_by_email` to find using the `email_username` as well for inbound emails with that as the TO address.

#### Note

This is safe to merge without impacting anyone seriously. If people had SMTP enabled for a group they would have IMAP enabled too currently, and that is a very small amount of users because IMAP is an alpha product, and also because the UserNotification change has a guard to make sure it is not used if IMAP is enabled for the group. The existing IMAP tests work, and I tested this functionality by manually POSTing replies to the SMTP address into my local discourse.

There will probably be more work needed on this, but it needs to be tested further in a real hosted environment to continue.

### CODE REVIEW

File: app/mailers/user_notifications.rb [hunk 1]
```diff
@@ -520,9 +528,48 @@ def send_notification_email(opts)
...
+       !group.imap_enabled &&
+       SiteSetting.enable_smtp &&
+       opts[:use_group_smtp_if_configured]
+
+      port, enable_tls, enable_starttls_auto = EmailSettingsValidator.provider_specific_ssl_overrides(
+        group.smtp_server, group.smtp_port, group.smtp_ssl, group.smtp_ssl
+      )
+
+      # TODO (martin): Remove this once testing is over and this is more stable.
+      Rails.logger.warn("Using SMTP settings from group #{group.name} (#{group.id}) to send user notification for topic #{post.topic.id} and user #{user.id} (#{user.email})")
```
Comment (SamSaffron) [inline app/mailers/user_notifications.rb:550]: mixed on merging it in with this warning, I guess if we really want this then: 

if ENV['DISCOURSE_VERBOSE_GROUP_MESSAGES']
   ...
end

Get rid of that block later on

**Review by SamSaffron — APPROVED** [review]

**Comment by SamSaffron** [comment]
looks pretty safe to merge to me.

File: app/mailers/user_notifications.rb (hunk 1 above)
Comment (martin-brennan) [reply app/mailers/user_notifications.rb:550]: I don't intend to leave this in there once it goes to Meta/other places, this is mostly for my benefit when I do some testing on our hosted site. I really doubt anyone has this setting enabled right now for it to show in our logs

Comment (SamSaffron) [reply app/mailers/user_notifications.rb:550]: OK, this is very minor, go ahead and merge as is.

**Pull request merged**